mempool/clist_mempool.go

```diff
@@ -701,38 +774,45 @@ func (mem *CListMempool) recheckTxs() {
 	if n := mem.recheck.numPendingTxs.Load(); n > 0 {
 		mem.logger.Error("Not all txs were rechecked", "not-rechecked", n)
 	}
-	mem.logger.Debug("Done rechecking txs", "height", mem.height.Load(), "num-txs", mem.Size())
+
+	mem.logger.Debug("Done rechecking", "height", mem.height.Load(), "num-txs", mem.Size())
 }
 
-// The cursor and end pointers define a dynamic list of transactions that could be rechecked. The
-// end pointer is fixed. When a recheck response for a transaction is received, cursor will point to
-// the entry in the mempool corresponding to that transaction, thus narrowing the list. Transactions
-// corresponding to entries between the old and current positions of cursor will be ignored for
-// rechecking. This is to guarantee that recheck responses are processed in the same sequential
-// order as they appear in the mempool.
+// When a recheck response for a transaction is received, cursor will point to
+// the entry in the mempool corresponding to that transaction, advancing the
+// cursor, thus narrowing the list of transactions to recheck. In case there are
+// entries between the previous and the current positions of cursor, they will
+// be ignored for rechecking. This is to guarantee that recheck responses are
+// processed in the same sequential order as they appear in the mempool.
 type recheck struct {
-	cursor        *clist.CElement // next expected recheck response
-	end           *clist.CElement // last entry in the mempool to recheck
-	doneCh        chan struct{}   // to signal that rechecking has finished successfully (for async app connections)
-	numPendingTxs atomic.Int32    // number of transactions still pending to recheck
-	isRechecking  atomic.Bool     // true iff the rechecking process has begun and is not yet finished
-	recheckFull   atomic.Bool     // whether rechecking TXs cannot be completed before a new block is decided
+	iter          *NonBlockingIterator
+	cursor        Entry         // next expected recheck response
+	doneCh        chan struct{} // to signal that rechecking has finished successfully (for async app connections)
+	numPendingTxs atomic.Int32  // number of transactions still pending to recheck
+	isRechecking  atomic.Bool   // true iff the rechecking process has begun and is not yet finished
+	recheckFull   atomic.Bool   // whether rechecking TXs cannot be completed before a new block is decided
 }
 
-func newRecheck() *recheck {
-	return &recheck{
-		doneCh: make(chan struct{}, 1),
-	}
+func newRecheck(iter *NonBlockingIterator) *recheck {
+	r := recheck{}
+	r.iter = iter
+	return &r
 }
 
-func (rc *recheck) init(first, last *clist.CElement) {
+func (rc *recheck) init(mp *CListMempool) {
```

Comment:
Two minor comments:
* Shouldn't it be more _compact_ if it was `newRecheck` the one that received `mp`, rather than `init`?
  * At the end of the day, the recheck is designed to always work with the same mempool, isn't it?
* As a follow-up, if `newRecheck` is now receiving `mp`, then it doesn't need to get `iter`, it can create it just as `init` does

Reply:
I added the parameter for `newRecheck` but I do not see a way around moving the argument from `init` without having a `mp` field within the `recheck`struct?

Reply:
Yes, that's what I was thinking... rather than providing the mempool at every call, just provide it a creation time (`newRecheck`) and then reuse for any further call. For that, you need to keep a reference to the mempool in the struct itself. Do you see a problem with that?

Reply:
No just double checking

Reply:
Done @sergio-mena

Reply:
🙏